pkg/mapper/fsm/fsm.go

```diff
@@ -0,0 +1,375 @@
+	minRemainingLength int
+	maxRemainingLength int
+	// result* members are nil unless there's a metric ends with this state
+	result                interface{}
+	resultPriority        int
+	resultNameFormatter   *templateFormatter
+	resultLabelsFormatter map[string]*templateFormatter
+}
+
+type fsmBacktrackStackCursor struct {
+	fieldIndex     int
+	captureIndex   int
+	currentCapture string
+	state          *mappingState
+	prev           *fsmBacktrackStackCursor
+	next           *fsmBacktrackStackCursor
+}
+
+type FSM struct {
+	root              *mappingState
+	needsBacktracking bool
+	metricTypes       []string
+	disableOrdering   bool
+	statesCount       int
+}
+
+// NewFSM creates a new FSM instance
+func NewFSM(metricTypes []string, maxPossibleTransitions int, disableOrdering bool) *FSM {
+	fsm := FSM{}
+	root := &mappingState{}
+	root.transitions = make(map[string]*mappingState, 3)
+
+	metricTypes = append(metricTypes, "")
+	for _, field := range metricTypes {
+		state := &mappingState{}
+		(*state).transitions = make(map[string]*mappingState, maxPossibleTransitions)
+		root.transitions[string(field)] = state
+	}
+	fsm.disableOrdering = disableOrdering
+	fsm.metricTypes = metricTypes
+	fsm.statesCount = 0
+	fsm.root = root
+	return &fsm
+}
+
+// AddState adds a state into the existing FSM
+func (f *FSM) AddState(match string, name string, labels prometheus.Labels, matchMetricType string,
+	maxPossibleTransitions int, result interface{}) {
+	// first split by "."
+	matchFields := strings.Split(match, ".")
+	// fill into our FSM
+	roots := []*mappingState{}
+	if matchMetricType == "" {
+		// if metricType not specified, connect the state from all three types
+		for _, metricType := range f.metricTypes {
+			roots = append(roots, f.root.transitions[string(metricType)])
+		}
+	} else {
+		roots = append(roots, f.root.transitions[matchMetricType])
+	}
+	var captureCount int
+	var finalStates []*mappingState
+	for _, root := range roots {
+		captureCount = 0
+		for i, field := range matchFields {
+			state, prs := root.transitions[field]
+			if !prs {
+				state = &mappingState{}
+				(*state).transitions = make(map[string]*mappingState, maxPossibleTransitions)
+				(*state).maxRemainingLength = len(matchFields) - i - 1
+				(*state).minRemainingLength = len(matchFields) - i - 1
+				root.transitions[field] = state
+				// if this is last field, set result to currentMapping instance
+				if i == len(matchFields)-1 {
+					root.transitions[field].result = result
+				}
+			} else {
+				(*state).maxRemainingLength = max(len(matchFields)-i-1, (*state).maxRemainingLength)
+				(*state).minRemainingLength = min(len(matchFields)-i-1, (*state).minRemainingLength)
+			}
+			if field == "*" {
+				captureCount++
+			}
+
+			// goto next state
+			root = state
+		}
+		finalStates = append(finalStates, root)
+	}
+	nameFmt := newTemplateFormatter(name, captureCount)
+
+	currentLabelFormatter := make(map[string]*templateFormatter, captureCount)
```

Comment:
it feels awkward that the FSM logic needs to know about "labels" and "names" … would it make sense to promote the whole formatting business into the mapper package, and make the API of the fsm package such that it can do that? then from the view of the FSM, this would all be subsumed under the "result" field.

Reply:
yeah it makes more sense, and we don't need to pass those variables around. let me try to implement that.